### PR TITLE
f_aws-network-firewall-rule-group Add IP Prefix list

### PR DESCRIPTION
**Pull request by meetreks**
### Description

The change introduces adding IP Prefix list to the rule group. Further details can be found in this link - https://docs.aws.amazon.com/network-firewall/latest/APIReference/API_CreateRuleGroup.html

### Relations

Relates to https://github.com/hashicorp/terraform-provider-aws/issues/27421
Closes #0000

### References

https://aws.amazon.com/blogs/networking-and-content-delivery/introducing-prefix-lists-in-aws-network-firewall-stateful-rule-groups/

### Output from Acceptance Testing

```
Test Evidence:
:~/terraform-provider-aws (f-aws_networkfirewall_rule_group) $ make testacc PKG=networkfirewall TESTS=TestAccNetworkFirewallRuleGroup_Basic_referenceSets
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/networkfirewall/... -v -count 1 -parallel 20 -run='TestAccNetworkFirewallRuleGroup_Basic_referenceSets' -timeout 180m
=== RUN TestAccNetworkFirewallRuleGroup_Basic_referenceSets
=== PAUSE TestAccNetworkFirewallRuleGroup_Basic_referenceSets
=== CONT TestAccNetworkFirewallRuleGroup_Basic_referenceSets
--- PASS: TestAccNetworkFirewallRuleGroup_Basic_referenceSets (168.00s)
PASS
ok [github.com/hashicorp/terraform-provider-aws/internal/service/networkfirewall](http://github.com/hashicorp/terraform-provider-aws/internal/service/networkfirewall) (http://github.com/hashicorp/terraform-provider-aws/internal/service/networkfirewall) 168.142s
:~/terraform-provider-aws (f-aws_networkfirewall_rule_group) $

Update Function Test Results
:~/terraform-provider-aws (f-aws_networkfirewall_rule_group) $ make testacc PKG=networkfirewall TESTS=TestAccNetworkFirewallRuleGroup_Basic_updateReferenceSets
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/networkfirewall/... -v -count 1 -parallel 20 -run='TestAccNetworkFirewallRuleGroup_Basic_updateReferenceSets'  -timeout 180m
=== RUN   TestAccNetworkFirewallRuleGroup_Basic_updateReferenceSets
=== PAUSE TestAccNetworkFirewallRuleGroup_Basic_updateReferenceSets
=== CONT  TestAccNetworkFirewallRuleGroup_Basic_updateReferenceSets
--- PASS: TestAccNetworkFirewallRuleGroup_Basic_updateReferenceSets (185.71s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/networkfirewall    185.848s
:~/terraform-provider-aws (f-aws_networkfirewall_rule_group) $ 

...
```

### CODE REVIEW

File: internal/service/networkfirewall/rule_group.go [hunk 1]
```diff
@@ -58,6 +58,46 @@ func ResourceRuleGroup() *schema.Resource {
...
+												"ip_set_reference": {
+													Type:     schema.TypeList,
+													Required: true,
+													MaxItems: 5,
+													Elem: &schema.Resource{
+														Schema: map[string]*schema.Schema{
+															"reference_arn": {
+																Type:     schema.TypeString,
+																Required: true,
+																//Elem:     &schema.Schema{Type: schema.TypeString},
```
Comment (silvaalbert) [inline internal/service/networkfirewall/rule_group.go:90]: Please remove this comment, and add validation
```suggestion
				ValidateFunc: verify.ValidARN,
```

File: internal/service/networkfirewall/rule_group_test.go [hunk 2]
```diff
@@ -56,6 +56,41 @@ func TestAccNetworkFirewallRuleGroup_Basic_rulesSourceList(t *testing.T) {
...
+					testAccCheckRuleGroupExists(resourceName, &ruleGroup),
+					acctest.CheckResourceAttrRegionalARN(resourceName, "arn", "network-firewall", fmt.Sprintf("stateful-rulegroup/%s", rName)),
+					resource.TestCheckResourceAttr(resourceName, "capacity", "100"),
+					resource.TestCheckResourceAttr(resourceName, "name", rName),
+					resource.TestCheckResourceAttr(resourceName, "type", networkfirewall.RuleGroupTypeStateful),
+					resource.TestCheckResourceAttr(resourceName, "rule_group.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "rule_group.0.reference_sets.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "rule_group.0.reference_sets.0.ip_set_references.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "rule_group.0.reference_sets.0.ip_set_references.0.ip_set_reference.#", "1"),
+					//resource.TestCheckTypeSetElemAttr(resourceName, "rule_group.0.reference_sets.0.ip_set_references.*.ip_set_reference.0.reference_arn.*", "aws_ec2_managed_prefix_list.this.arn"),
```
Comment (silvaalbert) [inline internal/service/networkfirewall/rule_group_test.go:82]: Please remove this comment
```suggestion

```

File: internal/service/networkfirewall/rule_group_test.go [hunk 3]
```diff
@@ -56,6 +56,41 @@ func TestAccNetworkFirewallRuleGroup_Basic_rulesSourceList(t *testing.T) {
 	})
 }
 
+func TestAccNetworkFirewallRuleGroup_Basic_referenceSets(t *testing.T) {
```
Comment (silvaalbert) [inline internal/service/networkfirewall/rule_group_test.go:59]: Please consider adding an 'update' test to ensure the resource is updated properly

**Review by silvaalbert — CHANGES_REQUESTED** [review]
Looking good, I provided a couple of comments for your consideration

File: internal/service/networkfirewall/rule_group.go (hunk 1 above)
Comment (meetreks) [reply internal/service/networkfirewall/rule_group.go:90]: Changed as requested.

File: internal/service/networkfirewall/rule_group_test.go (hunk 3 above)
Comment (meetreks) [reply internal/service/networkfirewall/rule_group_test.go:59]: Added Update Test as requested.

**Review by AdamTylerLynch — CHANGES_REQUESTED** [review]
A few changes requested.

File: internal/service/networkfirewall/rule_group.go [hunk 4]
```diff
@@ -58,6 +58,46 @@ func ResourceRuleGroup() *schema.Resource {
...
+													ValidateFunc: validation.All(
+														validation.StringLenBetween(1, 32),
+														validation.StringMatch(regexp.MustCompile(`^[A-Za-z]`), "must begin with alphabetic character"),
+														validation.StringMatch(regexp.MustCompile(`^[A-Za-z0-9_]+$`), "must contain only alphanumeric and underscore characters"),
+													),
+												},
+												"ip_set_reference": {
+													Type:     schema.TypeList,
+													Required: true,
+													MaxItems: 5,
```
[the remark this answers is not in the document]
Comment (meetreks) [reply internal/service/networkfirewall/rule_group.go:84]: In the Console it allows 5 Keys and 5 Reference ARN(s) and that is now I know it takes 5.

File: internal/service/networkfirewall/rule_group_test.go [hunk 5]
```diff
@@ -56,6 +56,87 @@ func TestAccNetworkFirewallRuleGroup_Basic_rulesSourceList(t *testing.T) {
...
```
[the remark this answers is not in the document]
Comment (meetreks) [reply internal/service/networkfirewall/rule_group_test.go:59]: Made changes as suggested and reran accpt tests.

File: internal/service/networkfirewall/rule_group.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (meetreks) [reply internal/service/networkfirewall/rule_group.go]: Removed MaxItems:5 as suggested and letting the API to throw errors if greater than 5 items added via HCL

**Review by AdamTylerLynch — APPROVED** [review]
LGTM 🚀

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

```console
% make testacc TESTARGS='-run=TestAccNetworkFirewallRuleGroup_' PKG=networkfirewall ACCTEST_PARALLELISM=3
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/networkfirewall/... -v -count 1 -parallel 3  -run=TestAccNetworkFirewallRuleGroup_ -timeout 180m
=== RUN   TestAccNetworkFirewallRuleGroup_Basic_rulesSourceList
=== PAUSE TestAccNetworkFirewallRuleGroup_Basic_rulesSourceList
=== RUN   TestAccNetworkFirewallRuleGroup_Basic_referenceSets
=== PAUSE TestAccNetworkFirewallRuleGroup_Basic_referenceSets
=== RUN   TestAccNetworkFirewallRuleGroup_Basic_updateReferenceSets
=== PAUSE TestAccNetworkFirewallRuleGroup_Basic_updateReferenceSets
=== RUN   TestAccNetworkFirewallRuleGroup_Basic_statefulRule
=== PAUSE TestAccNetworkFirewallRuleGroup_Basic_statefulRule
=== RUN   TestAccNetworkFirewallRuleGroup_Basic_statelessRule
=== PAUSE TestAccNetworkFirewallRuleGroup_Basic_statelessRule
=== RUN   TestAccNetworkFirewallRuleGroup_Basic_rules
=== PAUSE TestAccNetworkFirewallRuleGroup_Basic_rules
=== RUN   TestAccNetworkFirewallRuleGroup_statefulRuleOptions
=== PAUSE TestAccNetworkFirewallRuleGroup_statefulRuleOptions
=== RUN   TestAccNetworkFirewallRuleGroup_updateStatefulRuleOptions
=== PAUSE TestAccNetworkFirewallRuleGroup_updateStatefulRuleOptions
=== RUN   TestAccNetworkFirewallRuleGroup_statelessRuleWithCustomAction
=== PAUSE TestAccNetworkFirewallRuleGroup_statelessRuleWithCustomAction
=== RUN   TestAccNetworkFirewallRuleGroup_updateRules
=== PAUSE TestAccNetworkFirewallRuleGroup_updateRules
=== RUN   TestAccNetworkFirewallRuleGroup_updateRulesSourceList
=== PAUSE TestAccNetworkFirewallRuleGroup_updateRulesSourceList
=== RUN   TestAccNetworkFirewallRuleGroup_rulesSourceAndRuleVariables
=== PAUSE TestAccNetworkFirewallRuleGroup_rulesSourceAndRuleVariables
=== RUN   TestAccNetworkFirewallRuleGroup_updateStatefulRule
=== PAUSE TestAccNetworkFirewallRuleGroup_updateStatefulRule
=== RUN   TestAccNetworkFirewallRuleGroup_updateMultipleStatefulRules
=== PAUSE TestAccNetworkFirewallRuleGroup_updateMultipleStatefulRules
=== RUN   TestAccNetworkFirewallRuleGroup_StatefulRule_action
=== PAUSE TestAccNetworkFirewallRuleGroup_StatefulRule_action
=== RUN   TestAccNetworkFirewallRuleGroup_StatefulRule_header
=== PAUSE TestAccNetworkFirewallRuleGroup_StatefulRule_header
=== RUN   TestAccNetworkFirewallRuleGroup_updateStatelessRule
=== PAUSE TestAccNetworkFirewallRuleGroup_updateStatelessRule
=== RUN   TestAccNetworkFirewallRuleGroup_tags
=== PAUSE TestAccNetworkFirewallRuleGroup_tags
=== RUN   TestAccNetworkFirewallRuleGroup_disappears
=== PAUSE TestAccNetworkFirewallRuleGroup_disappears
=== CONT  TestAccNetworkFirewallRuleGroup_Basic_rulesSourceList
=== CONT  TestAccNetworkFirewallRuleGroup_disappears
=== CONT  TestAccNetworkFirewallRuleGroup_updateRules
--- PASS: TestAccNetworkFirewallRuleGroup_Basic_rulesSourceList (143.40s)
=== CONT  TestAccNetworkFirewallRuleGroup_Basic_rules
--- PASS: TestAccNetworkFirewallRuleGroup_disappears (145.03s)
=== CONT  TestAccNetworkFirewallRuleGroup_statelessRuleWithCustomAction
--- PASS: TestAccNetworkFirewallRuleGroup_updateRules (166.15s)
=== CONT  TestAccNetworkFirewallRuleGroup_updateStatefulRuleOptions
--- PASS: TestAccNetworkFirewallRuleGroup_Basic_rules (139.39s)
=== CONT  TestAccNetworkFirewallRuleGroup_statefulRuleOptions
--- PASS: TestAccNetworkFirewallRuleGroup_statelessRuleWithCustomAction (138.15s)
=== CONT  TestAccNetworkFirewallRuleGroup_StatefulRule_action
--- PASS: TestAccNetworkFirewallRuleGroup_statefulRuleOptions (140.37s)
=== CONT  TestAccNetworkFirewallRuleGroup_tags
--- PASS: TestAccNetworkFirewallRuleGroup_updateStatefulRuleOptions (284.38s)
=== CONT  TestAccNetworkFirewallRuleGroup_updateStatelessRule
--- PASS: TestAccNetworkFirewallRuleGroup_StatefulRule_action (196.84s)
=== CONT  TestAccNetworkFirewallRuleGroup_StatefulRule_header
--- PASS: TestAccNetworkFirewallRuleGroup_tags (148.77s)
=== CONT  TestAccNetworkFirewallRuleGroup_Basic_statefulRule
--- PASS: TestAccNetworkFirewallRuleGroup_updateStatelessRule (185.12s)
=== CONT  TestAccNetworkFirewallRuleGroup_Basic_statelessRule
--- PASS: TestAccNetworkFirewallRuleGroup_StatefulRule_header (172.80s)
=== CONT  TestAccNetworkFirewallRuleGroup_Basic_updateReferenceSets
--- PASS: TestAccNetworkFirewallRuleGroup_Basic_statefulRule (132.89s)
=== CONT  TestAccNetworkFirewallRuleGroup_updateStatefulRule
--- PASS: TestAccNetworkFirewallRuleGroup_Basic_statelessRule (163.33s)
=== CONT  TestAccNetworkFirewallRuleGroup_updateMultipleStatefulRules
--- PASS: TestAccNetworkFirewallRuleGroup_Basic_updateReferenceSets (187.58s)
=== CONT  TestAccNetworkFirewallRuleGroup_rulesSourceAndRuleVariables
--- PASS: TestAccNetworkFirewallRuleGroup_updateStatefulRule (154.65s)
=== CONT  TestAccNetworkFirewallRuleGroup_updateRulesSourceList
--- PASS: TestAccNetworkFirewallRuleGroup_updateMultipleStatefulRules (168.56s)
=== CONT  TestAccNetworkFirewallRuleGroup_Basic_referenceSets
--- PASS: TestAccNetworkFirewallRuleGroup_rulesSourceAndRuleVariables (155.01s)
--- PASS: TestAccNetworkFirewallRuleGroup_updateRulesSourceList (165.04s)
--- PASS: TestAccNetworkFirewallRuleGroup_Basic_referenceSets (144.48s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/networkfirewall	1119.876s
```

**Comment by ewbankkit** [comment]
@meetreks Thanks for the contribution :tada: :clap:.

**Pull request merged**